### PR TITLE
home-assistant: Add matter support

### PR DESCRIPTION
**Pull request by mweinelt**
###### Description of changes

Matter support in home-assistant.

https://www.home-assistant.io/integrations/matter/

Need to investigate whether the module needs some hardening relaxed. Feedback welcome!

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by dotlambda — COMMENTED** [review]
Please set `meta.sourceProvenance`.

**Comment by mweinelt** [comment]
Two packages are providing parts to the `chip`module, and there are imports from one into the other. The packages are

- home-assistant-chip-core
- home-assistant-chip-clusters

In the test phase of `python-matter-server` the `home-assistant-chip-core` package tries to import bits that should be provided by `home-assistant-chip-clusters`, but it fails to find them.

In `python-matter-server` the `home-assistant-chip-clusters` package is propagated, and the `home-assistant-chip-core` package is optional and passed into `checkInputs`.

```
_________________ ERROR collecting tests/server/test_server.py _________________
ImportError while importing test module '/build/source/tests/server/test_server.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/nix/store/al6g1zbk8li6p8mcyp0h60d08jaahf8c-python3-3.10.9/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/server/test_server.py:8: in <module>
    from chip.clusters import ClusterCommand
/nix/store/92n7kjlhvda17lnvnal47alnq4y5gw8y-python3.10-home-assistant-chip-core-2022.12.0/lib/python3.10/site-packages/chip/clusters/__init__.py:24: in <module>
    from . import Command
/nix/store/92n7kjlhvda17lnvnal47alnq4y5gw8y-python3.10-home-assistant-chip-core-2022.12.0/lib/python3.10/site-packages/chip/clusters/Command.py:26: in <module>
    from .ClusterObjects import ClusterCommand
E   ModuleNotFoundError: No module named 'chip.clusters.ClusterObjects'
=========================== short test summary info ============================
ERROR tests/server/test_server.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
```

When I create a nix-shell using `python3.withPackages (ps: with ps; [ home-assistant-chip-core home-assistant-chip-clusters ])` I can do the import just fine.

```
❯ nix-shell -I nixpkgs=$(pwd) -p "python3.withPackages (ps: with ps; [ home-assistant-chip-clusters home-assistant-chip-core ])"
this derivation will be built:
  /nix/store/j6hk7yk3vhkn7201ansy0ys7834zrd67-python3-3.10.9-env.drv
building '/nix/store/j6hk7yk3vhkn7201ansy0ys7834zrd67-python3-3.10.9-env.drv'...
created 466 symlinks in user environment
❯ python3
Python 3.10.9 (main, Dec  6 2022, 18:44:57) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from chip.clusters import ClusterCommand
>>> 
```

Why does it fail during check phase?

File: pkgs/development/python-modules/mobly/default.nix [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  ];
+
+  checkInputs = [
+    procps
+    pytestCheckHook
+    pytz
+  ];
+
+  meta = with lib; {
+    changelog = "";
```
Comment (dotlambda) [inline pkgs/development/python-modules/mobly/default.nix:45]:
```suggestion
    changelog = "https://github.com/google/mobly/blob/${src.rev}/CHANGELOG.md";
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by mweinelt** [comment]
Well, I assume that's because for the `checkPhase` I get a `PYTHONPATH` cobbled together, while for `python3.withPackages` I get an actual shared site-packages directory, where everything is linked together.

With `PYTHONPATH` the lookup seems saturated as soon as it finds its first `chip` package.

File: pkgs/development/python-modules/python-matter-server/default.nix [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    pytestCheckHook
+  ]
+  ++ lib.flatten (builtins.attrValues passthru.optional-dependencies);
+
+  preCheck = let
+    pythonEnv = python.withPackages (_: propagatedBuildInputs ++ checkInputs);
+  in
+  ''
+    export PYTHONPATH=${pythonEnv}/${python.sitePackages}
+  '';
```
Comment (mweinelt) [inline pkgs/development/python-modules/python-matter-server/default.nix:72]: @dotlambda WDYT about this hack to make the tests work?

File: pkgs/development/python-modules/python-matter-server/default.nix [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  ];
+
+  passthru.optional-dependencies = {
+    server = [
+      home-assistant-chip-core
+    ];
+  };
+
+  checkInputs = [
+    pytest
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/python-matter-server/default.nix:61]:
```suggestion
```
already included in pytestCheckHook

File: pkgs/development/python-modules/home-assistant-chip-core/default.nix [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+let
+  wheels = {
+    "aarch64-linux" = fetchurl {
+      url = "https://files.pythonhosted.org/packages/38/04/a5886ac7356d1202ec6586bf01e24201fefc52d10f54795229558b81f039/home_assistant_chip_core-2022.12.0-cp37-abi3-manylinux_2_31_aarch64.whl";
+      hash = "sha256-oNqrvbzXeXpMG3v9RK6kppONH4n7xLVaJCEFXxVj2jE=";
+    };
+    "x86_64-linux" = fetchurl {
+      url = "https://files.pythonhosted.org/packages/8b/3a/5c68a504407188c3b6b183fc2958b4fe5aafe17ebc0c442f1e3587c090c7/home_assistant_chip_core-2022.12.0-cp37-abi3-manylinux_2_31_x86_64.whl";
+      hash = "sha256-S5n1MUig8ZDSLgWeVmu+5qLZ4kfHQUC9qZcVfM8rPvw=";
+    };
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/home-assistant-chip-core/default.nix:34]: just that it is not forgotten in the chat history

https://files.pythonhosted.org/packages/cp37/h/home_assistant_chip_core/home_assistant_chip_core-2022.12.0-cp37-abi3-manylinux_2_31_x86_64.whl

File: pkgs/development/python-modules/python-matter-server/default.nix (hunk 3 above)
Comment (mweinelt) [reply pkgs/development/python-modules/python-matter-server/default.nix:61]: Remove it and give it a try.

File: pkgs/development/python-modules/home-assistant-chip-core/default.nix (hunk 4 above)
Comment (mweinelt) [reply pkgs/development/python-modules/home-assistant-chip-core/default.nix]: Wondering if this is an improvement at all.

[`140fd27` (#210297)](https://github.com/NixOS/nixpkgs/pull/210297/commits/140fd2715d1787af3db0a8cb5d76950ff0c7fb0d)

**Comment by FRidh** [comment]
> Well, I assume that's because for the `checkPhase` I get a `PYTHONPATH` cobbled together, while for `python3.withPackages` I get an actual shared site-packages directory, where everything is linked together.
> 
> With `PYTHONPATH` the lookup seems saturated as soon as it finds its first `chip` package.

You are in the right direction.

With old-fashioned namespace packages we could not "merge" an environment using `PYTHONPATH`; it would stop at the first package and you would not have the chance to load the other ones. With PEP 420 namespace packages this is not the case, and you can have them on `PYTHONPATH`.

Checking the wheels, `home-assistant-chip-clusters` does not have a `__init__.py` and is thus a PEP 420 namespace package. `home-assistant-chip-core` on the other hand does have a (empty) `__init__.py` and registers as a regular package. At that point, it won't be able to load `home-assistant-chip-clusters` anymore. Thus, we have a packaging mistake here. The `__init__.py` from `home-assistant-chip-chip` needs to be removed so it registers as a PEP 420 namespace package as well.

**Comment by mweinelt** [comment]
I tried setting `pythonNamespaces` to get rid of `__init__py`s that would prevent PEP420 compat, but nested module lookups into `chip.clusters` would still fail for one of the packages.

```
# core
lib/python3.10/site-packages/chip/clusters/
lib/python3.10/site-packages/chip/clusters/Attribute.py
lib/python3.10/site-packages/chip/clusters/Command.py
lib/python3.10/site-packages/chip/clusters/__pycache__/
lib/python3.10/site-packages/chip/clusters/__pycache__/Attribute.cpython-310.pyc
lib/python3.10/site-packages/chip/clusters/__pycache__/Command.cpython-310.pyc

# clusters
lib/python3.10/site-packages/chip/clusters/
lib/python3.10/site-packages/chip/clusters/CHIPClusters.py
lib/python3.10/site-packages/chip/clusters/ClusterObjects.py
lib/python3.10/site-packages/chip/clusters/Objects.py
lib/python3.10/site-packages/chip/clusters/TestObjects.py
lib/python3.10/site-packages/chip/clusters/Types.py
lib/python3.10/site-packages/chip/clusters/__pycache__/
lib/python3.10/site-packages/chip/clusters/__pycache__/CHIPClusters.cpython-310.pyc
lib/python3.10/site-packages/chip/clusters/__pycache__/ClusterObjects.cpython-310.pyc
lib/python3.10/site-packages/chip/clusters/__pycache__/Objects.cpython-310.pyc
lib/python3.10/site-packages/chip/clusters/__pycache__/TestObjects.cpython-310.pyc
lib/python3.10/site-packages/chip/clusters/__pycache__/Types.cpython-310.pyc
```

```pytb
_________________ ERROR collecting tests/server/test_server.py _________________
ImportError while importing test module '/build/source/tests/server/test_server.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/nix/store/al6g1zbk8li6p8mcyp0h60d08jaahf8c-python3-3.10.9/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/server/test_server.py:8: in <module>
    from chip.clusters import ClusterCommand
E   ImportError: cannot import name 'ClusterCommand' from 'chip.clusters' (unknown location)
=========================== short test summary info ============================
ERROR tests/server/test_server.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.18s ===============================
```

That's because the `__init__.py` below `chip.clusters` isn't boilerplate, but includes the required imports.

```python
"""Provides Python APIs for CHIP."""
from . import Command
from . import Attribute
from .Objects import *
from . import CHIPClusters
```

This means we can't remove it and as such can't achieve PEP420 compatible namespacing between the two packages.

This warrants an upstream issue, but should not block the inclusion of matter support in our home-assistant as it is.

**Comment by mweinelt** [comment]
Result of `nixpkgs-review pr 210297` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>home-assistant-component-tests.matter</li>
    <li>python-matter-server (python310Packages.python-matter-server)</li>
    <li>python310Packages.home-assistant-chip-clusters</li>
    <li>python310Packages.home-assistant-chip-core</li>
    <li>python310Packages.mobly</li>
    <li>python39Packages.home-assistant-chip-clusters</li>
    <li>python39Packages.home-assistant-chip-core</li>
    <li>python39Packages.mobly</li>
    <li>python39Packages.python-matter-server</li>
  </ul>
</details>

**Pull request merged**